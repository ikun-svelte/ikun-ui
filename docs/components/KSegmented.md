---
title: KSegmented
lang: en-US
---

# KSegmented

When displaying multiple options and user can select a single option.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/segmented
```

```bash [yarn]
yarn add @ikun-ui/segmented
```

```bash [npm]
npm install @ikun-ui/segmented
```

:::

## Basic usage

The most basic usage.

<demo src="segmented/basic.svelte"  github='Segmented'></demo>

## Block segmented

`block` property will make the Segmented fit to its parent width.

<demo src="segmented/block.svelte"  github='Segmented'></demo>

## Disabled

Disabled Segmented.

<demo src="segmented/disabled.svelte"  github='Segmented'></demo>

## Custom render

Custom each Segmented Item by slot.

<demo src="segmented/custom.svelte"  github='Segmented'></demo>

## Sizes

There are three sizes of an Segmented

<demo src="segmented/size.svelte"  github='Segmented'></demo>

## Icon

Set `icon` for Segmented Item.

<demo src="segmented/icon.svelte"  github='Segmented'></demo>

## Just Only icon

Set `icon` without `label` for Segmented Item.

<demo src="segmented/only.svelte"  github='Segmented'></demo>

## Segmented Props

| Name  | Type                     | Default | Description                                |
| ----- | ------------------------ | ------- | ------------------------------------------ |
| value | `string \| number`       | `''`    | Selected value.                            |
| size  | `sm \| md \| lg`         | `md`    | Size of segmented                          |
| block | `boolean`                | `false` | Option to fit width to its parent's width. |
| cls   | `string`                 | `''`    | Additional class for component             |
| attrs | `Record<string, string>` | `{}`    | Additional attributes                      |

## Segmented Events

| Name   | Description                                                     | Type                               |
| ------ | --------------------------------------------------------------- | ---------------------------------- |
| change | The callback function that is triggered when the state changes. | `(value: string \| number)=> void` |

## Segmented Slots

| Name    | Description                    |
| ------- | ------------------------------ |
| default | content of segmented component |

## SegmentedItem Props

| Name     | Type                     | Default | Description                                                                                                     |
| -------- | ------------------------ | ------- | --------------------------------------------------------------------------------------------------------------- |
| value    | `string \| number`       | `-`     | The callback value of `SegmentedItem`. Clicking on `SegmentedItem` will be passed to the event as a parameter.. |
| disabled | `boolean`                | `false` | Whether the `SegmentedItem` is disabled.                                                                        |
| label    | `string`                 | `-`     | Display value of `SegmentedItem`.                                                                               |
| icon     | `string`                 | `-`     | The front icon of `SegmentedItem`, which should comply with unocss standard.                                    |
| onlyIcon | `boolean`                | `false` | Show only icons.                                                                                                |
| cls      | `string`                 | `''`    | Additional class for component                                                                                  |
| attrs    | `Record<string, string>` | `{}`    | Additional attributes                                                                                           |

## SegmentedItem Events

| Name  | Description                                                       | Type                               |
| ----- | ----------------------------------------------------------------- | ---------------------------------- |
| click | The callback function that is triggered when click segmentedItem. | `(value: string \| number)=> void` |

## SegmentedItem Slots

| Name    | Description                        |
| ------- | ---------------------------------- |
| default | content of segmentedItem component |
