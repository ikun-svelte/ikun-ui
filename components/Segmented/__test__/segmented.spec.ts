import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import KSegmented from '../src';
import KSegmentedValue from './fixture/value.svelte';
import KSegmentedBlock from './fixture/block.svelte';
import KSegmentedSize from './fixture/size.svelte';
import KSegmentedLabel from './fixture/label.svelte';
import KSegmentedIcon from './fixture/icon.svelte';
import KSegmentedOnly from './fixture/only.svelte';
import KSegmentedClick from './fixture/click.svelte';
import KSegmentedDisabled from './fixture/disabled.svelte';
import KSegmentedCustom from './fixture/custom.svelte';
import { tick } from 'svelte';
let host;

const initHost = () => {
	host = globalThis.document.createElement('div');
	host.setAttribute('id', 'host');
	globalThis.document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.useRealTimers();
});

describe('Test: KSegmented', () => {
	test('props: cls', async () => {
		//@ts-ignore
		const instance = new KSegmented({
			target: host,
			props: {
				cls: 'k-segmented--test'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-segmented--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: value', async () => {
		//@ts-ignore
		const instance = new KSegmentedValue({
			target: host
		});
		expect(instance).toBeTruthy();
		const valElm = host.querySelector('#test_value');
		expect(valElm.innerHTML.includes('Daily-value')).toBeTruthy();

		const items = host.querySelectorAll('.k-segmented-item');
		items[2].click();
		await tick();
		expect(valElm.innerHTML.includes('Monthly-value')).toBeTruthy();
	});

	test('props: block', async () => {
		//@ts-ignore
		const instance = new KSegmentedBlock({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size', async () => {
		//@ts-ignore
		const instance = new KSegmentedSize({
			target: host
		});
		expect(instance).toBeTruthy();
		const lgBtn = host.querySelector('#test_lg');
		lgBtn.click();
		await tick();
		expect(host.innerHTML).matchSnapshot();

		const mdBtn = host.querySelector('#test_md');
		mdBtn.click();
		await tick();
		expect(host.innerHTML).matchSnapshot();

		const smBtn = host.querySelector('#test_sm');
		smBtn.click();
		await tick();
		expect(host.innerHTML).matchSnapshot();
	});

	test('events: change', async () => {
		//@ts-ignore
		const instance = new KSegmentedValue({
			target: host
		});
		expect(instance).toBeTruthy();
		const valElm = host.querySelector('#test_value');
		expect(valElm.innerHTML.includes('Daily-value')).toBeTruthy();

		const items = host.querySelectorAll('.k-segmented-item');
		items[2].click();
		await tick();
		expect(valElm.innerHTML.includes('Monthly-value')).toBeTruthy();
	});
});

describe('Test: KSegmentedItem', () => {
	test('props: value', async () => {
		//@ts-ignore
		const instance = new KSegmentedValue({
			target: host
		});
		expect(instance).toBeTruthy();
		const items = host.querySelectorAll('.k-segmented-item');
		expect(items[0].innerHTML.includes('k-segmented-item--active')).toBeTruthy();

		items[2].click();
		await tick();
		expect(items[2].innerHTML.includes('k-segmented-item--active')).toBeTruthy();
	});

	test('props: block', async () => {
		//@ts-ignore
		const instance = new KSegmentedLabel({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: icon', async () => {
		//@ts-ignore
		const instance = new KSegmentedIcon({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: onlyIcon', async () => {
		//@ts-ignore
		const instance = new KSegmentedOnly({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		//@ts-ignore
		const instance = new KSegmentedDisabled({
			target: host
		});
		expect(instance).toBeTruthy();
		const valElm = host.querySelector('#test_value');
		expect(valElm.innerHTML.includes('Daily-value')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();

		const items = host.querySelectorAll('.k-segmented-item');
		items[2].click();
		await tick();
		expect(valElm.innerHTML.includes('Daily-value')).toBeTruthy();
	});

	test('slots: default', async () => {
		//@ts-ignore
		const instance = new KSegmentedCustom({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('events: click', async () => {
		//@ts-ignore
		const instance = new KSegmentedClick({
			target: host
		});
		expect(instance).toBeTruthy();
		const valElm = host.querySelector('#test_value');
		expect(valElm.innerHTML.includes('Daily-value')).toBeTruthy();

		const items = host.querySelectorAll('.k-segmented-item');
		items[2].click();
		await tick();
		expect(valElm.innerHTML.includes('Monthly-value')).toBeTruthy();
	});
});
