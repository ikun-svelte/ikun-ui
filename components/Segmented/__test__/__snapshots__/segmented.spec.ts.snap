// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KSegmented > props: block 1`] = `"<div class="k-segmented k-segmented__dark k-segmented-block"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item k-segmented-item--block" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item k-segmented-item--block" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item k-segmented-item--block" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item k-segmented-item--block" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item k-segmented-item--block" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div>"`;

exports[`Test: KSegmented > props: cls 1`] = `"<div class="k-segmented k-segmented__dark k-segmented--test"><div class="k-segmented-group"></div></div>"`;

exports[`Test: KSegmented > props: size 1`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--lg k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--lg k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--lg k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--lg k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--lg k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <button id="test_lg">Daily-value</button> <button id="test_md">Daily-value</button> <button id="test_sm">Daily-value</button>"`;

exports[`Test: KSegmented > props: size 2`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <button id="test_lg">Daily-value</button> <button id="test_md">Daily-value</button> <button id="test_sm">Daily-value</button>"`;

exports[`Test: KSegmented > props: size 3`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--sm k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--sm k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--sm k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--sm k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--sm k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <button id="test_lg">Daily-value</button> <button id="test_md">Daily-value</button> <button id="test_sm">Daily-value</button>"`;

exports[`Test: KSegmentedItem > props: block 1`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <div id="test_value">Daily-value</div>"`;

exports[`Test: KSegmentedItem > props: disabled 1`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item--disabled" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <div id="test_value">Daily-value</div>"`;

exports[`Test: KSegmentedItem > props: icon 1`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span><span class="k-icon--base k-icon--base__dark k-icon-transition i-icon-logo-vue k-segmented-item--icon" role="" aria-hidden="true" style="width: 18px; height: 18px;"></span> Daily</span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <div id="test_value">Daily-value</div>"`;

exports[`Test: KSegmentedItem > props: onlyIcon 1`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Daily" class="k-segmented-item--label"><span><span class="k-icon--base k-icon--base__dark k-icon-transition i-icon-logo-vue k-segmented-item--icon" role="" aria-hidden="true" style="width: 18px; height: 18px;"></span> </span></div> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Weekly" class="k-segmented-item--label"><span> Weekly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Monthly" class="k-segmented-item--label"><span> Monthly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Quarterly" class="k-segmented-item--label"><span> Quarterly</span></div> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><div title="Yearly" class="k-segmented-item--label"><span> Yearly</span></div> </div></div></div> <div id="test_value">Daily-value</div>"`;

exports[`Test: KSegmentedItem > slots: default 1`] = `"<div class="k-segmented k-segmented__dark"><div class="k-segmented-group"><div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><span class="custom-segmented-item svelte-1lybzi">Daily 🐔</span> <div class="k-segmented-item--active k-segmented-item--active__dark"></div></div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><span class="custom-segmented-item svelte-1lybzi">Weekly 🎤</span> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><span class="custom-segmented-item svelte-1lybzi">Monthly 💃</span> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><span class="custom-segmented-item svelte-1lybzi">Quarterly ✌️</span> </div> <div class="k-segmented-item--md k-segmented-item__dark k-segmented-item" aria-hidden="true"><span class="custom-segmented-item svelte-1lybzi">Yearly 🏀</span> </div></div></div>"`;
